heartbeat/monitors/active/icmp/loop.go

```diff
@@ -92,13 +94,25 @@ func newICMPLoop() (*icmpLoop, error) {
 	// IPv4/IPv6 checking
 	conn4 := createListener("IPv4", "ip4:icmp")
 	conn6 := createListener("IPv6", "ip6:ipv6-icmp")
-
+	unprivilegedPossible := false
 	l := &icmpLoop{
 		conn4:    conn4,
 		conn6:    conn6,
 		recv:     make(chan packet, 16),
 		requests: map[requestID]*requestContext{},
 	}
+	if conn4 == nil && conn6 == nil {
+
+		switch runtime.GOOS {
+		case "darwin":
+		case "linux":
+			unprivilegedPossible = true
+			conn4 = createListener("Unprivileged IPv4", "udp4")
+			conn6 = createListener("Unprivileged IPv6", "udp6")
+			l.conn4 = conn4
+			l.conn6 = conn6
+		}
+	}
 
 	if conn4 != nil {
 		go l.runICMPRecv(conn4, protocolICMP)
```

Comment:
Perhaps here we should not pass `l.conn4`? Perhaps @urso remembers?

Reply:
For raw sockets (icmp solution) we need to pass the socket, as the worker loop generates and parses the actual ICMP packets. The socket is not really connection based, so it is more similar to a raw packet sniffer. I think this should be fine with the UDP based approach as well, but not really necessary.

The icmpLoop is a singleton, as for raw ICMP we should need only one actual socket. It stores conn4/conn6 in `l.conn4` only, so we can close the socket on shutdown. We pass the socket, cause we create two loops, one for IPv4 and one for IPv6. Alternatively we have to write 2 separate worker loops.

Reply:
It is needed for socket disposing/close